In [1]:
#importing essential libraries

import numpy as np
import pandas as pd

# importing Preprocessing libraries
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# importing Regression Algorithms
from sklearn.linear_model import LinearRegression, Ridge, Lasso

# importing libraries for calculating Metrics
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [2]:
# loading the dataset

df = pd.read_csv('energydata_complete.csv')
df.head(3)

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  

[3 rows x 29 columns]

In [3]:
# dropping the 'date' and 'lights' columns

df = df.drop(['date', 'lights'], axis=1)
df.head(3)

Appliances     T1       RH_1    T2       RH_2     T3       RH_3         T4  \
0          60  19.89  47.596667  19.2  44.790000  19.79  44.730000  19.000000   
1          60  19.89  46.693333  19.2  44.722500  19.79  44.790000  19.000000   
2          50  19.89  46.300000  19.2  44.626667  19.79  44.933333  18.926667   

        RH_4         T5  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  45.566667  17.166667  ...  17.033333  45.53  6.600000        733.5    92.0   
1  45.992500  17.166667  ...  17.066667  45.56  6.483333        733.6    92.0   
2  45.890000  17.166667  ...  17.000000  45.50  6.366667        733.7    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  
2   6.333333   55.333333        5.1  28.642668  28.642668  

[3 rows x 27 columns]

In [4]:
# Normalising the dataset

scaler = MinMaxScaler()
norm_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [5]:
norm_df.head(3)

Appliances       T1      RH_1        T2      RH_2        T3      RH_3  \
0    0.046729  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066   
1    0.046729  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871   
2    0.037383  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569   

         T4      RH_4        T5  ...        T9      RH_9     T_out  \
0  0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1  0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369239   
2  0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365488   

   Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0     0.097674  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1     0.100000  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2     0.102326  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  

[3 rows x 27 columns]

## QUESTION 12
#### From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two D.P?

In [6]:
x = df['T2']
y = df['T6']

In [7]:
x

0        19.200000
1        19.200000
2        19.200000
3        19.200000
4        19.200000
           ...    
19730    25.890000
19731    25.754000
19732    25.628571
19733    25.414000
19734    25.264286
Name: T2, Length: 19735, dtype: float64

In [8]:
y

0         7.026667
1         6.833333
2         6.560000
3         6.433333
4         6.366667
           ...    
19730    24.796667
19731    24.196667
19732    23.626667
19733    22.433333
19734    21.026667
Name: T6, Length: 19735, dtype: float64

In [9]:
# reshaping 'x' so as to let us fit with the model

array = np.array(x)
x_feature = array.reshape(-1,1)

In [10]:
# splitting to training and testing set

x_train, x_test, y_train, y_test = train_test_split(x_feature, y, test_size=0.3, random_state=42)

In [11]:
# training with Linear regression algorithm

linear_model = LinearRegression()
linear_model.fit(x_train, y_train)
pred_values = linear_model.predict(x_test)

In [12]:
# determining the R^2 value

r2_score = r2_score(y_test, pred_values)
round(r2_score, 2)

0.64

##### The R^2 Value is 0.64

## QUESTION 13
#### What is the Mean Absolute Error (in two decimal places)?

In [13]:
features = norm_df.drop('Appliances', axis=1)
target = norm_df['Appliances']

In [14]:
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42)

In [15]:
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((13814, 26), (5921, 26), (13814,), (5921,))

In [16]:
model = LinearRegression()
model.fit(x_train, y_train)
preds = model.predict(x_test)

In [17]:
# getting the mean absolute error

mae = mean_absolute_error(y_test, preds)
round(mae, 2)

0.05

##### The MAE is 0.05

## QUESTION 14
#### What is the Residual Sum of Squares (in two decimal places)?

In [18]:
# residual sum of squares

rss = np.sum(np.square(y_test - preds))
round(rss, 2)

45.35

##### The RSS is 45.35

## QUESTION 15
#### What is the Root Mean Squared Error (in three decimal places)?

In [19]:
# root mean squared error

rmse = np.sqrt(mean_squared_error(y_test, preds))
round(rmse, 3)

0.088

##### The RMSE is 0.088

## QUESTION 16
#### What is the Coefficient of Determination (in two decimal places)?

In [20]:
# coefficient of determination (also known as r2_score)
from sklearn.metrics import r2_score
r2 = r2_score(y_test, preds)
round(r2, 2)

0.15

##### The Coefficient of Determination is 0.15

## QUESTION 17
#### Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [21]:
# creating a function for getting the weights

def determine_weights_df(model, feat, col_name):
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

In [22]:
model_weights = determine_weights_df(model, x_train, 'Linear_Model_Weight')
model_weights

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

##### The features with the lowest and highest weights are RH_2 and RH_1 respectively

## QUESTION 18
#### Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [23]:
# Training the Ridge regression model

R_model = Ridge(alpha = 0.4)
R_model.fit(x_train, y_train)

Ridge(alpha=0.4)

In [24]:
R_preds = R_model.predict(x_test)

In [25]:
# finding the root mean squared error

R_rmse = np.sqrt(mean_squared_error(y_test, R_preds))
round(R_rmse, 3)

0.088

##### The value of RMSE of the Ridge model is 0.088, which is the same as the linear model. 
##### Therefore, there is no change.

## QUESTION 19
#### Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [26]:
# Training the lasso regressor

lasso_model = Lasso(alpha= 0.001)
lasso_model.fit(x_train, y_train)

Lasso(alpha=0.001)

In [27]:
lasso_preds = lasso_model.predict(x_test)

In [28]:
lasso_weights = determine_weights_df(lasso_model, x_train, 'Lasso_weight')
lasso_weights

Features  Lasso_weight
0        RH_out     -0.049557
1          RH_8     -0.000110
2            T1      0.000000
3     Tdewpoint      0.000000
4    Visibility      0.000000
5   Press_mm_hg     -0.000000
6         T_out      0.000000
7          RH_9     -0.000000
8            T9     -0.000000
9            T8      0.000000
10         RH_7     -0.000000
11          rv1     -0.000000
12           T7     -0.000000
13           T6      0.000000
14         RH_5      0.000000
15           T5     -0.000000
16         RH_4      0.000000
17           T4     -0.000000
18         RH_3      0.000000
19           T3      0.000000
20         RH_2     -0.000000
21           T2      0.000000
22         RH_6     -0.000000
23          rv2     -0.000000
24    Windspeed      0.002912
25         RH_1      0.017880

In [29]:
non_zero_feature_weight = np.sum(lasso_weights != 0)
non_zero_feature_weight

Features        26
Lasso_weight     4
dtype: int64

##### From the results above, there are 4 features that have non-zero feature weights

## QUESTION 20
#### What is the new RMSE with the Lasso Regression (in 3 decimal places)?

In [30]:
L_rmse = np.sqrt(mean_squared_error(y_test, lasso_preds))

In [31]:
round(L_rmse, 3)

0.094

##### The new RMSE value is 0.094